Modules/_testcapi/unicode.c

```diff
@@ -659,6 +675,78 @@ unicode_getdefaultencoding(PyObject *self, PyObject *Py_UNUSED(ignored))
     return PyBytes_FromString(s);
 }
 
+/* Test PyUnicode_Decode() */
+static PyObject *
+unicode_decode(PyObject *self, PyObject *args)
```

Comment:
We are now converting test for C APIs to use AC
https://github.com/python/cpython/issues/104469

Would you like to apply it here too?

General guide for this: https://github.com/python/cpython/issues/104469#issuecomment-1547730882

Reply:
I have doubts about these changes.

1. It makes the code less transparent. For C API tests we should be sure that the parsing code does not introduce some side effects. For example, it can change the reference counts, or raise additional exceptions.
2. It affects the core interpreter by adding interned strings.
3. C tests serve as implicit tests for `PyArg_Parse*()` functions.

Reply:
> 1. It makes the code less transparent. For C API tests we should be sure that the parsing code does not introduce some side effects. For example, it can change the reference counts, or raise additional exceptions.

I don't think so, I think that we should believe in the implementation of AC for the future(see also https://github.com/faster-cpython/ideas/issues/565 and https://github.com/python/cpython/pull/104384#issuecomment-1543818134, and if AC makes any side effects we should fix the AC, not the test code.
IMO, Using AC should not care where to use it.


> 2. It affects the core interpreter by adding interned strings.

I think that we can avoid the issue by the rule `Try to avoid adding new interned strings; reuse existing parameter names if possible and use the as feature to override the C name`

> 3. C tests serve as implicit tests for PyArg_Parse*() functions.

Can we add a separate test for `PyArg_Parse*()` itself from the separation of concerns view?


> I intentionally did not use AC in the C API tests.

Well, I will not touch the Unicode C APIs if you want. But I would like to suggest focusing on C API tests themselves Manual parsing is not an important part of test codes but maintenance should be easier for anyone.

Reply:
@corona10 @serhiy-storchaka:

1. we have dedicated tests for Argument Clinic; if side effects such as ref leaks are introduced, we'll know; if bugs appear, we fix them and add new clinic tests
2. that's a bit unfair IMO; none of the recent changes introduced new interned strings, and it is an issue we pay attention to
3. there are dedicated tests for `PyArg_Parse*` in `_testcapi`; if coverage is too low, we should expand on those tests

Also, I'd put it the other way around:

A lot of the test code has badly written boilerplate code. Very often, there is missing error handling, potential ref leaks, etc. Getting rid of boilerplate code is a good thing, IMO.

Reply:
I meant what if Argument Clinic keeps a reference to an argument it can affect testing the code which depends on whether the refcount == 1 or > 1. There is no such case in this PR, but currently I can see what exactly parsing code does. With Argument Clinic the parsing code will be in other file. If it suddenly changes I would not know whether it affects testing or not. Maybe I'm worrying for nothing.

The condition of my eyes only allows me to see a small part of the screen at a time. In this PR all functions fit in 10 lines, so I can see the parsing code and the code that does the work. Argument Clinic would insert several lines between argument declarations and the actual code, it is too much for me. I will not be able to see it as a whole. But I understand that other people can take it easier.

@corona10, @erlend-aasland, could you create a new PR based on this PR, but with using Argument Clinic? It's hard for me and I'm sure I'll make a few mistakes.

Reply:
> I meant what if Argument Clinic keeps a reference to an argument it can affect testing the code which depends on whether the refcount == 1 or > 1.

Ah, yes that's a corner case to consider. Thanks for the heads-up.

If needed, we can apply clinic in a follow-up PR. Let's not block this PR. Adding the tests more important IMO.